### PR TITLE
[RST-2427] Added a 'source' field to the constraints

### PR DESCRIPTION
**Pull request by svwilliams**
Added a 'source' field to the constraints. This is an API-breaking change and hence touches a lot of files.

### CODE REVIEW

**Review by ayrton04 — APPROVED** [review]

File: fuse_constraints/include/fuse_constraints/absolute_constraint_impl.h [hunk 1]
```diff
@@ -48,10 +48,11 @@ namespace fuse_constraints
 
 template<class Variable>
 AbsoluteConstraint<Variable>::AbsoluteConstraint(
+  const std::string& source,
   const Variable& variable,
   const fuse_core::VectorXd& mean,
   const fuse_core::MatrixXd& covariance) :
-    fuse_core::Constraint{variable.uuid()},
+    fuse_core::Constraint(source, {variable.uuid()}),  // NOLINT(whitespace/braces)
```
Comment (efernandez) [inline fuse_constraints/include/fuse_constraints/absolute_constraint_impl.h:55]: I think this should also work: `fuse_core::Constraint{source, variable.uuid()},` (same in other instances below)

File: fuse_constraints/include/fuse_constraints/relative_constraint_impl.h [hunk 2]
```diff
@@ -48,11 +48,12 @@ namespace fuse_constraints
 
 template<class Variable>
 RelativeConstraint<Variable>::RelativeConstraint(
+  const std::string& source,
   const Variable& variable1,
   const Variable& variable2,
   const fuse_core::VectorXd& delta,
   const fuse_core::MatrixXd& covariance) :
-    fuse_core::Constraint{variable1.uuid(), variable2.uuid()},
+    fuse_core::Constraint(source, {variable1.uuid(), variable2.uuid()}),  // NOLINT(whitespace/braces)
```
Comment (efernandez) [inline fuse_constraints/include/fuse_constraints/relative_constraint_impl.h:56]: As mentioned before, I think this should work: `fuse_core::Constraint{source, variable1.uuid(), variable2.uuid()},` (same below)

File: fuse_constraints/src/absolute_orientation_3d_stamped_constraint.cpp [hunk 3]
```diff
@@ -40,33 +40,38 @@
...
 namespace fuse_constraints
 {
 
 AbsoluteOrientation3DStampedConstraint::AbsoluteOrientation3DStampedConstraint(
+  const std::string& source,
   const fuse_variables::Orientation3DStamped& orientation,
   const fuse_core::Vector4d& mean,
   const fuse_core::Matrix3d& covariance) :
-    fuse_core::Constraint{orientation.uuid()},
+    fuse_core::Constraint(source, {orientation.uuid()}),  // NOLINT(whitespace/braces)
```
Comment (efernandez) [inline fuse_constraints/src/absolute_orientation_3d_stamped_constraint.cpp:54]: `fuse_core::Constraint{source, orientation.uuid()},`

**Review by efernandez — APPROVED** [review]
LGTM, but it seems to me that you're missing direct initialization (https://en.cppreference.com/w/cpp/language/direct_initialization) with list initialization (https://en.cppreference.com/w/cpp/language/list_initialization), or maybe aggregate initialization (https://en.cppreference.com/w/cpp/language/aggregate_initialization) :thinking:

File: fuse_constraints/include/fuse_constraints/absolute_constraint_impl.h (hunk 1 above)
Comment (svwilliams) [reply fuse_constraints/include/fuse_constraints/absolute_constraint_impl.h:55]: That particular incantation results in:
```
error: no matching function for call to ‘std::vector<boost::uuids::uuid>::vector(boost::uuids::uuid&)’
```
This does work, but I'm not sure it has any advantages over the constructor `()` version:
```
fuse_core::Constraint{source, {variable.uuid()}}
```
I could do something with variadic templates:
https://stackoverflow.com/questions/25721869/initializer-list-combined-with-other-parameters
```
public:
  template <class ...Args>
  Constraint(const std::string& source, Args... args) : Constraint(source, {args...}) {}
private:
  Constraint(const std::string& source, std::initializer_list<UUID> variable_uuid_list) :
    source_(source),
    variables_(variable_uuid_list)
  {}
```
but I'm not sure it is worth the trouble.

Comment (efernandez) [reply fuse_constraints/include/fuse_constraints/absolute_constraint_impl.h:55]: Fair enough. I don't think it's worth the trouble, and I didn't realized there's an `std::initializer_list` for the UUIDs.

**Pull request merged**